### PR TITLE
Fix significant join performance regression

### LINKED ISSUE #37489: 10x Performance drop on join queries on 21.4 & 2.15
### Setup
```
AWS: c5d.4xlarge

CREATE TEMPORARY TABLE _v AS
SELECT
    xxHash64(rand()) % 100000 AS _did,
    toString(number + (1 * 10000)) AS _pid,
    toString(number + (1 * 1000)) AS _aid,
    toString(number + (1 * 1000)) AS _sid,
    toDate(parseDateTimeBestEffort('2022-01-01') + (rand() % 2600000)) AS v_date
FROM numbers(1000000)

CREATE TEMPORARY TABLE _i AS
SELECT
    xxHash64(rand()) % 40000 AS _did,
    toDate(i_datestamp) AS i_date,
    parseDateTimeBestEffort('2022-01-15') - (rand() % 2600000) AS i_datestamp,
    toString((rand() % 99) + 1) AS ad__aid,
    toString((rand() % 99) + 1) AS ad__sid,
    xxHash32(number) % 2000 AS _dim_hash,
    rand() % 10 AS i_count
FROM numbers(40000000)

```

## Query
```
SELECT * FROM ( SELECT * FROM _i ) INNER JOIN (SELECT _did, v_date FROM _v ) USING (_did) WHERE v_date > i_datestamp FORMAT `Null`
```

## Result for various clickhouse versions

```

22.5.1.2079
0 rows in set. Elapsed: 23.645 sec. Processed 41.00 million rows, 1.48 GB (1.73 million rows/s., 62.54 MB/s.)

22.4.5.9 
0 rows in set. Elapsed: 23.577 sec. Processed 41.00 million rows, 1.48 GB (1.74 million rows/s., 62.72 MB/s.)

------

22.3.6.5
0 rows in set. Elapsed: 1.817 sec. Processed 41.00 million rows, 1.48 GB (22.56 million rows/s., 813.64 MB/s.)


22.2.3.5 
0 rows in set. Elapsed: 1.799 sec. Processed 41.00 million rows, 1.48 GB (22.79 million rows/s., 821.93 MB/s.)

22.1.4.30
0 rows in set. Elapsed: 2.966 sec. Processed 41.00 million rows, 1.32 GB (13.82 million rows/s., 444.66 MB/s.)

21.12.4.1
0 rows in set. Elapsed: 1.901 sec. Processed 41.00 million rows, 1.48 GB (21.57 million rows/s., 777.91 MB/s.)

21.11.11.1 
0 rows in set. Elapsed: 1.912 sec. Processed 41.00 million rows, 1.48 GB (21.44 million rows/s., 773.22 MB/s.)

21.10.5.3
0 rows in set. Elapsed: 1.992 sec. Processed 41.00 million rows, 1.48 GB (20.58 million rows/s., 742.30 MB/s.)

21.8.13.6 (our current production)
0 rows in set. Elapsed: 2.067 sec. Processed 41.00 million rows, 1.48 GB (19.84 million rows/s., 715.48 MB/s.)
```

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix significant join performance regression which was introduced in https://github.com/ClickHouse/ClickHouse/pull/35616 . It's interesting that common join queries such as ssb queries have been 10 times slower for almost 3 months while no one complains.

Close https://github.com/ClickHouse/ClickHouse/issues/37489

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Interpreters/HashJoin.cpp [hunk 1]
```diff
@@ -1020,12 +1018,12 @@ class AddedColumns
         {
             for (size_t j = 0, size = right_indexes.size(); j < size; ++j)
             {
-                auto column_from_block = block.getByPosition(right_indexes[j]);
```
Comment (amosbird) [inline src/Interpreters/HashJoin.cpp:1023]: This line of code is the regression root cause. We should not do any heavy operation (block copy) inside hot loop.

Comment (antonio2368) [reply src/Interpreters/HashJoin.cpp:1023]: Thanks for fixing and explaining the issue!
I didn't know about the `insertFromFullColumn` but it's really similar to what we do with `nullable` columns so I should've expected for something like that to exist.

File: src/Interpreters/HashJoin.cpp [hunk 2]
```diff
@@ -1003,15 +1003,13 @@ class AddedColumns
...
-                {
-                    JoinCommon::changeLowCardinalityInplace(column_from_block);
-                }
-
+                const auto & column_from_block = block.getByPosition(right_indexes[j]);
                 if (auto * nullable_col = typeid_cast<ColumnNullable *>(columns[j].get());
                     nullable_col && !column_from_block.column->isNullable())
                     nullable_col->insertFromNotNullable(*column_from_block.column, row_num);
+                else if (auto * lowcard_col = typeid_cast<ColumnLowCardinality *>(columns[j].get());
+                         lowcard_col && !typeid_cast<const ColumnLowCardinality *>(column_from_block.column.get()))
```
Comment (antonio2368) [inline src/Interpreters/HashJoin.cpp:1011]: why do we check if it's not `const ColumnLowCardinality *`?

**Comment by vdimir** [comment]
Is it related to https://github.com/ClickHouse/ClickHouse/issues/37489 ?

**Comment by amosbird** [comment]
Ah, so there is indeed a complain)

**Review by antonio2368 — APPROVED** [review]

**Comment by antonio2368** [comment]
@Mergifyio update

**Pull request merged**